Android/app/src/main/java/app/intra/net/doh/Race.java

```diff
@@ -32,39 +35,45 @@
     void onResult(int index);
   }
 
-  private final List<Probe> probes;
-  private final Listener listener;
-
   /**
-   * Creates a race between different servers.  To run the race, call start().
-   * @param factory The factory will be called once for each server.
+   * Starts a race between different servers.
+   * @param context Used to read the IP addresses of the servers from storage.
    * @param urls The URLs for all the DOH servers to compare.
    * @param listener Called once on an arbitrary thread with the result of the race.
    */
-  public Race(ServerConnectionFactory factory, String[] urls, Listener listener) {
-    probes = new ArrayList<>(urls.length);
-    for (int i = 0; i < urls.length; ++i) {
-      probes.add(new Probe(factory, urls[i], new Callback(i)));
-    }
-    this.listener = listener;
+  public static void start(Context context, String[] urls, Listener listener) {
+    Probe.Factory factory = RemoteConfig.getUseGoDoh() ? GoProbe.factory : JavaProbe.factory;
+    start(factory, context, urls, listener);
   }
 
-  public synchronized void start() {
-    for (Probe task : probes) {
-      task.start();
+  // Exposed for unit testing only.
+  static void start(Probe.Factory factory, Context context, String[] urls, Listener listener) {
+    List<Probe> probes = new ArrayList<>(urls.length);
+    for (int i = 0; i < urls.length; ++i) {
+      probes.add(factory.get(context, urls[i], new Callback(i, probes, listener)));
```

Comment:
It doesn't seem like you need a factory. You could simply pass a function that takes a URL and a done(error) callback and runs the test. No need to create an object for each URL.

The use of protected methods is usually a Code Health red flag. The functionality you are reusing via inheritance can be here instead, and the Probes can simply be functions. Or a simple class with a single test/run method as described above.

Reply:
Done.  This is much simpler.

This design no longer supports cancelling outstanding probes, but probes are not expensive so cancelling them is not very important, and in the previous implementation cancellation was only partially supported.  (Cancellation was only possible before the probe query is sent).